### PR TITLE
Free-for-all pre-0.1 fixathon

### PR DESCRIPTION
**Pull request by nullstyle**
This Gigantic PR contains a series of changes in an effort to reach a taggable 0.1 release of horizon.
## `resource` package

This biggest change is the introduction of the `resource` package that provides a home for all the JSON responses that get rendered in horizon.  This PR significantly changes how we populate JSON, introducing the convention of having a `Populate` method that is provided with backing data to populate the JSON struct with.  The way in which we handle "polymorphic" JSON responses, such as operations and effects is also changed.  A `Base` struct provides the common fields and each shape of the resource has a separate struct that wraps `Base`, allowing us to be strongly typed for all of our JSON responses.
## Homogenize action methods

This PR also attempts to bring every "Action" method to parity with regards to coding conventions.  Since much of horizon's helper infrastructure was built as we build features, many of the older actions did not take advantage of newer helpers.  
## Upgrades to `hal` package: Replace halgo and hal.Page

halgo is no longer used to populate links in our JSON responses. `hal.Page` was replaced with a smarter implementation that can automatically populate links and has a nicer API for adding page records.
## Grab bag fixes
- Setting the `Err` field on an action will now automatically close a streamed response after transmitting the error to the client.
- #194 is fixed
- Lots of code formatting and ordering changes
- Upgraded stellar base to latest
- With the update to `stellar-base`, horizon now scans database values to xdr datatypes when possible.
- A timeout response is now sent instead of simply killing the request with no response.

### CODE REVIEW

File: src/github.com/stellar/horizon/db/asset.go [hunk 1]
```diff
@@ -7,8 +7,8 @@ import (
 
 // This file contains helpers for working with stellar assets
 
-func assetFromDB(typ int32, code string, issuer string) (result xdr.Asset, err error) {
-	switch xdr.AssetType(typ) {
+func assetFromDB(typ xdr.AssetType, code string, issuer string) (result xdr.Asset, err error) {
+	switch typ {
```
Comment (bartekn) [inline src/github.com/stellar/horizon/db/asset.go:11]: Is it a typo or it has to be `typ` for some reason?

Comment (nullstyle) [inline src/github.com/stellar/horizon/db/asset.go:11]: `type` is a reserved word in go.  I've been using `typ` and `aType` randomly to work around it.

Comment (bartekn) [inline src/github.com/stellar/horizon/db/asset.go:11]: OK!

**Comment by bartekn** [comment]
LGTM :+1: 

> A timeout response is now sent instead of simply killing the request with no response.

:sparkles:

**Pull request merged**